### PR TITLE
Restructure adb error classes.

### PR DESCRIPTION
**Pull request by xpconanfan**
* AdbTimeoutError should not be a subclass of AdbError as they are
  raised for different occasions.
* Improve log messages.
* Add proper docstrings.

@oprypin

---
This change is [Reviewable](https://reviewable.io/reviews/google/mobly/368)

### CODE REVIEW

File: mobly/controllers/android_device_lib/adb.py [hunk 1]
```diff
@@ -121,27 +139,27 @@ def _exec_cmd(self, args, shell, timeout):
             AdbError: The adb command exit code is not 0.
             AdbTimeoutError: The adb command timed out.
         """
+        cmd_str = ' '.join(args)
```
Comment (oprypin) [inline mobly/controllers/android_device_lib/adb.py]: This changes the way command line arguments are displayed from `['cp', 'my file', 'other file']` to `"cp my file other file"`. The change is not strictly necessary for this pull request, but if you'd like to proceed with it, I think the proper way would be:

```python
def _quote_command(command):
  return ' '.join(pipes.quote(arg) for arg in command)
```
...returning `cp 'my file' 'other file'`

File: mobly/controllers/android_device_lib/adb.py [hunk 2]
```diff
@@ -121,27 +139,27 @@ def _exec_cmd(self, args, shell, timeout):
...
         (out, err) = proc.communicate()
         ret = proc.returncode
-        logging.debug('cmd: %s, stdout: %s, stderr: %s, ret: %s', args, out,
+        logging.debug('cmd: %s, stdout: %s, stderr: %s, ret: %s', cmd_str, out,
                       err, ret)
         if ret == 0:
             return out
         else:
-            raise AdbError(cmd=args, stdout=out, stderr=err, ret_code=ret)
+            raise AdbError(cmd=cmd_str, stdout=out, stderr=err, ret_code=ret)
```
Comment (oprypin) [inline mobly/controllers/android_device_lib/adb.py]: The `cmd` member should stay as it was - a list. Changing it to a string breaks API without a real need. I would suggest only changing the way that the string representation of the exception works (or not changing it)

File: mobly/controllers/android_device_lib/adb.py [hunk 3]
```diff
@@ -215,10 +233,7 @@ def instrument(self, package, options=None, runner=None):
         options_string = ' '.join(options_list)
 
         instrumentation_command = 'am instrument -r -w %s %s/%s' % (
-            options_string,
-            package,
-            runner,
-        )
+            options_string, package, runner, )
```
Comment (oprypin) [inline mobly/controllers/android_device_lib/adb.py]: This change with the comma at the end seems accidental

File: mobly/controllers/android_device_lib/adb.py [hunk 4]
```diff
@@ -31,8 +31,19 @@
...
-    """Raised when there is an error in adb operations."""
+class Error(Exception):
+    """Base error type for adb proxy module."""
+
+
+class AdbError(Error):
+    """Raised when an adb command encounters an error.
+
+    Args:
+        cmd: string, the adb command that timed out
```
Comment (oprypin) [inline mobly/controllers/android_device_lib/adb.py]: Should stay a list probably, in both places.

**Comment by oprypin** [comment]
Overall this is a good change. Thanks for it.

File: mobly/controllers/android_device_lib/adb.py (hunk 1 above)
Comment (xpconanfan) [reply mobly/controllers/android_device_lib/adb.py]: In debugging a failure caused by AdbError, I find it pretty common to try to run the same cmd manually to see the results. Having this as a list makes pasting the cmd difficult.

So the motivation here is so we can directly copy the cmd from the error msg.

`pipes.quote` is used for constructing cmd to be passed to pipes, whereas we purely want something for displaying. So I don't think `pipes` is necessary. Also `pipes` only works on POSIX, which is not good as we need to support windows as well.

Is there an alternative?

File: mobly/controllers/android_device_lib/adb.py (hunk 2 above)
Comment (xpconanfan) [reply mobly/controllers/android_device_lib/adb.py]: Done

File: mobly/controllers/android_device_lib/adb.py (hunk 3 above)
Comment (xpconanfan) [reply mobly/controllers/android_device_lib/adb.py]: Done.
Comma has always been there, just linter merged the lines..

File: mobly/controllers/android_device_lib/adb.py (hunk 4 above)
Comment (xpconanfan) [reply mobly/controllers/android_device_lib/adb.py]: Done

File: mobly/controllers/android_device_lib/adb.py (hunk 1 above)
Comment (oprypin) [reply mobly/controllers/android_device_lib/adb.py]: I have explained why correct quoting is necessary, with it you actually can copy/paste the command unmodified, without fearing for any unusual characters in the command. ` ` `;` `$` `*` `(` `)` `{` `}` `'` `"` - any of these will make the command invalid. And not just for the purposes of copy pasting, but even seeing what's going on.

I believe `pipes.quote` is available on Windows.
https://github.com/python/cpython/blob/2.7/Lib/pipes.py#L267
https://github.com/python/cpython/blob/3.6/Lib/pipes.py#L65

Comment (xpconanfan) [reply mobly/controllers/android_device_lib/adb.py]: Done

**Pull request merged**